pkg/controller/atlasproject/cloud_provider_access_test.go

```diff
@@ -26,30 +31,953 @@ func (c *cloudProviderAccessClient) GetRole(_ context.Context, _ string, _ strin
+	})
+
+	t.Run("should synchronize all operations without reach ready status", func(t *testing.T) {
+		cpas := []mdbv1.CloudProviderAccessRole{
+			{
+				ProviderName:      "AWS",
+				IamAssumedRoleArn: "aws:arn/my_role-1",
+			},
+			{
+				ProviderName:      "AWS",
+				IamAssumedRoleArn: "aws:arn/my_role-2",
+			},
+			{
+				ProviderName: "AWS",
+			},
+		}
+		atlasCPAs := []mongodbatlas.CloudProviderAccessRole{
+			{
+				AtlasAWSAccountARN:         "atlas-account-arn-1",
+				AtlasAssumedRoleExternalID: "atlas-external-role-id-1",
+				AuthorizedDate:             "authorized-date-1",
+				CreatedDate:                "created-date-1",
+				IAMAssumedRoleARN:          "aws:arn/my_role-1",
+				ProviderName:               "AWS",
+				RoleID:                     "role-1",
+			},
+			{
+				AtlasAWSAccountARN:         "atlas-account-arn-2",
+				AtlasAssumedRoleExternalID: "atlas-external-role-id-2",
+				CreatedDate:                "created-date-2",
+				IAMAssumedRoleARN:          "aws:arn/my_role-2",
+				ProviderName:               "AWS",
+				RoleID:                     "role-2",
+			},
+			{
+				AtlasAWSAccountARN:         "atlas-account-arn-4",
+				AtlasAssumedRoleExternalID: "atlas-external-role-id-4",
+				IAMAssumedRoleARN:          "aws:arn/my-role-4",
+				CreatedDate:                "created-date-4",
+				ProviderName:               "AWS",
+				RoleID:                     "role-4",
+			},
+		}
+		atlasClient := mongodbatlas.Client{
+			CloudProviderAccess: &cloudProviderAccessClient{
+				ListRolesFunc: func(projectID string) (*mongodbatlas.CloudProviderAccessRoles, *mongodbatlas.Response, error) {
+					return &mongodbatlas.CloudProviderAccessRoles{
+						AWSIAMRoles: atlasCPAs,
+					}, &mongodbatlas.Response{}, nil
+				},
+				CreateRoleFunc: func(projectID string, cpa *mongodbatlas.CloudProviderAccessRoleRequest) (*mongodbatlas.CloudProviderAccessRole, *mongodbatlas.Response, error) {
+					return &mongodbatlas.CloudProviderAccessRole{
+						AtlasAWSAccountARN:         "atlas-account-arn-3",
+						AtlasAssumedRoleExternalID: "atlas-external-role-id-3",
+						CreatedDate:                "created-date-3",
+						ProviderName:               "AWS",
+						RoleID:                     "role-3",
+					}, &mongodbatlas.Response{}, nil
+				},
+				AuthorizeRoleFunc: func(projectID, roleID string, cpa *mongodbatlas.CloudProviderAccessRoleRequest) (*mongodbatlas.CloudProviderAccessRole, *mongodbatlas.Response, error) {
+					atlasCPA := atlasCPAs[1]
+					atlasCPA.AuthorizedDate = "authorized-date-2"
+
+					return &atlasCPA, &mongodbatlas.Response{}, nil
+				},
+				DeauthorizeRoleFunc: func(cpa *mongodbatlas.CloudProviderDeauthorizationRequest) (*mongodbatlas.Response, error) {
+					return &mongodbatlas.Response{}, nil
+				},
+			},
+		}
+		workflowCtx := &workflow.Context{
+			Client: atlasClient,
+		}
+
+		result, err := syncCloudProviderAccess(context.TODO(), workflowCtx, "projectID", cpas)
+		assert.NoError(t, err)
```

Comment:
nit

Should this be `require.NoError(...)` instead? Or is the nil error also part of the final status expected in a way?

Reply:
As `require` fails the test immediately it's a bit annoying not knowing what went wrong in the first run. Given that, I think makes more sense to use it in places where we know, "if this doesn't work, everything else won't work and we should stop asap"